packages/core/src/lib/connection_manager.ts

```diff
@@ -0,0 +1,381 @@
+  ping: Libp2p["ping"];
+  dial: Libp2p["dial"];
+}
+
+export interface UpdatedStates {
+  allPeers: Peer[];
+  allDialablePeers: Peer[];
+  bootstrapPeers: Peer[];
+  nonBootstrapPeers: Peer[];
+  dialableBootstrapPeers: Peer[];
+  dialableNonBootstrapPeers: Peer[];
+  allConnections: Connection[];
+  bootstrapConnections: Connection[];
+  nonBootstrapConnections: Connection[];
+}
+
+export interface Options {
+  /**
+   * Number of attempts before a peer is considered non-dialable
+   * This is used to not spam a peer with dial attempts when it is not dialable
+   */
+  maxDialAttemptsForPeer?: number;
+  /**
+   * Max number of bootstrap peers allowed to be connected to, initially
+   * This is used to increase intention of dialing non-bootstrap peers, found using other discovery mechanisms (like Peer Exchange)
+   */
+  maxBootstrapPeersAllowed?: number;
+  // TODO: add this functionality back
+  // /**
+  //  * Number of attempts before dialling bootstrap nodes over the {@link maxBootstrapPeersAllowed} value.
+  //  * Dialing is the process of opening an outgoing connection to a listening peer with a transport that is supported by both peers.
+  //  * This is only used when other discovery mechanisms haven't yield peers that are dialable
+  //  * This increases relative centralization of the network
+  //  */
+  // maxTimeBeforeBootstrapFallbackMs?: number;
+  /**
+   * Interval at which to log the state of the connection manager
+   */
+  loggingIntervalMs?: number;
+}
+
+export class ConnectionManager {
+  private static instances = new Map<string, ConnectionManager>();
+  private keepAliveManager = new KeepAliveManager();
+  private options: Options;
+  private libp2pComponents: Libp2pComponents;
+  private dialAttemptsForPeer: Map<string, number> = new Map();
+
+  public static create(
+    peerId: string,
+    libp2p: Libp2pComponents,
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay,
+    options?: Options
+  ): ConnectionManager {
+    let instance = ConnectionManager.instances.get(peerId);
+    if (!instance)
+      instance = new ConnectionManager(
+        libp2p,
+        keepAliveOptions,
+        relay,
+        options
+      );
+    ConnectionManager.instances.set(peerId, instance);
+
+    return instance;
+  }
+
+  constructor(
+    libp2pComponents: Libp2pComponents,
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay,
+    options?: Options
+  ) {
+    this.libp2pComponents = libp2pComponents;
+    this.options = options ?? {};
+
+    this.runService(keepAliveOptions, relay)
+      .then(() => log(`Connection Manager is now running`))
+      .catch((error) => log(`Unexpected error while running service`, error));
+  }
+
+  private async runService(
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay
+  ): Promise<void> {
+    const { loggingIntervalMs = DEFAULT_LOGGING_INTERVAL_MS } = this.options;
+
+    // log state every N seconds
+    setInterval(async () => {
```

Comment:
I see that this function is run only from the constructor but I would still store this interval in the class instance to prevent from spawning many intervals. It would be good to clear this interval in the `stopService` too.